src/scope.cpp

```diff
@@ -0,0 +1,119 @@
+}
+
+template <typename Table, typename Entry>
+std::shared_ptr<Entry> ScopeStack::LookUpEntry_(
+    const std::string& id, std::unique_ptr<Table> Scope::*table) const {
+  ThrowIfNotInScope_();
+  // Iterates backward since we're using the container as a stack.
+  for (auto it = scopes_.crbegin(); it != scopes_.crend(); ++it) {
+    if (auto entry = ((*it).*table)->Probe(id)) {
+      return entry;
+    }
+  }
+  return nullptr;
+}
+
+template <typename Table, typename Entry>
+std::shared_ptr<Entry> ScopeStack::ProbeEntry_(
+    const std::string& id, std::unique_ptr<Table> Scope::*table) const {
+  ThrowIfNotInScope_();
+  return (scopes_.back().*table)->Probe(id);
+}
+
+std::shared_ptr<SymbolEntry> ScopeStack::AddSymbol(
+    std::unique_ptr<SymbolEntry> entry, ScopeKind kind) {
+  return AddEntry_<SymbolTable, SymbolEntry>(std::move(entry), kind,
+                                             &Scope::symbol_table);
+}
+
+std::shared_ptr<SymbolEntry> ScopeStack::LookUpSymbol(
+    const std::string& id) const {
+  return LookUpEntry_<SymbolTable, SymbolEntry>(id, &Scope::symbol_table);
+}
+
+std::shared_ptr<SymbolEntry> ScopeStack::ProbeSymbol(
+    const std::string& id) const {
+  return ProbeEntry_<SymbolTable, SymbolEntry>(id, &Scope::symbol_table);
+}
+
+std::shared_ptr<TypeEntry> ScopeStack::AddType(std::unique_ptr<TypeEntry> entry,
+                                               ScopeKind kind) {
+  return AddEntry_<TypeTable, TypeEntry>(std::move(entry), kind,
+                                         &Scope::type_table);
+}
+
+std::shared_ptr<TypeEntry> ScopeStack::LookUpType(const std::string& id) const {
+  return LookUpEntry_<TypeTable, TypeEntry>(id, &Scope::type_table);
+}
+
+std::shared_ptr<TypeEntry> ScopeStack::ProbeType(const std::string& id) const {
+  return ProbeEntry_<TypeTable, TypeEntry>(id, &Scope::type_table);
+}
+
+// Explicit template instantiation for the member functions used by ScopeStack
+template std::shared_ptr<SymbolEntry> ScopeStack::AddEntry_<
+    SymbolTable, SymbolEntry>(std::unique_ptr<SymbolEntry>, ScopeKind,
+                              std::unique_ptr<SymbolTable> Scope::*);
+
+template std::shared_ptr<TypeEntry> ScopeStack::AddEntry_<TypeTable, TypeEntry>(
+    std::unique_ptr<TypeEntry>, ScopeKind, std::unique_ptr<TypeTable> Scope::*);
+
+template std::shared_ptr<SymbolEntry>
+ScopeStack::LookUpEntry_<SymbolTable, SymbolEntry>(
+    const std::string&, std::unique_ptr<SymbolTable> Scope::*) const;
+
+template std::shared_ptr<TypeEntry>
+ScopeStack::LookUpEntry_<TypeTable, TypeEntry>(
+    const std::string&, std::unique_ptr<TypeTable> Scope::*) const;
+
+template std::shared_ptr<SymbolEntry>
+ScopeStack::ProbeEntry_<SymbolTable, SymbolEntry>(
+    const std::string&, std::unique_ptr<SymbolTable> Scope::*) const;
+
+template std::shared_ptr<TypeEntry>
+ScopeStack::ProbeEntry_<TypeTable, TypeEntry>(
+    const std::string&, std::unique_ptr<TypeTable> Scope::*) const;
```

Comment:
We don't need these explicit instantiations because they are already instantiated when defining the public functions (`AddSymbol`, `AddType`, etc.).